pkg/k8smgmt/ingress.go

```diff
@@ -0,0 +1,210 @@
+		pathType := networkingv1.PathTypePrefix
+		path.PathType = &pathType
+		path.Path = port.PathPrefix
+		if path.Path == "" {
+			path.Path = "/"
+		}
+		path.Backend.Service = &networkingv1.IngressServiceBackend{
+			Name: svcName,
+			Port: networkingv1.ServiceBackendPort{
+				Number: port.InternalPort,
+			},
+		}
+		httpRule.Paths = append(httpRule.Paths, path)
+	}
+	hostName := appInst.Uri
+	hostName = strings.TrimPrefix(hostName, "https://")
+	hostName = strings.TrimPrefix(hostName, "http://")
+	rule := networkingv1.IngressRule{}
+	rule.Host = hostName
+	rule.HTTP = &httpRule
+	ingress.Spec.Rules = []networkingv1.IngressRule{rule}
+	if hasTLS {
+		// Note the TLS secret is left blank to allow nginx to use
+		// the default SSL certificate.
+		tls := networkingv1.IngressTLS{
+			Hosts: []string{hostName},
+		}
+		ingress.Spec.TLS = append(ingress.Spec.TLS, tls)
+		ingress.ObjectMeta.Labels["nginx.ingress.kubernetes.io/ssl-redirect"] = "true"
+	}
+
+	// Apply the ingress spec
+	printer := &printers.YAMLPrinter{}
+	buf := bytes.Buffer{}
+	err = printer.PrintObj(&ingress, &buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal the ingress object to yaml, %s", err)
+	}
+	contents := buf.String()
+	kconfArg := names.GetTenantKconfArg()
+	fileName := getIngressManifestName(names)
+	err = pc.WriteFile(client, fileName, contents, "k8s ingress", pc.NoSudo)
+	if err != nil {
+		return nil, err
+	}
+	cmd := fmt.Sprintf("kubectl %s apply -f %s", kconfArg, fileName)
+	log.SpanLog(ctx, log.DebugLevelInfra, "applying ingress", "cmd", cmd)
+	out, err := client.Output(cmd)
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply ingress command %s: %s, %s", cmd, out, err)
+	}
+	log.SpanLog(ctx, log.DebugLevelInfra, "applied ingress")
+	return &ingress, nil
+}
+
+func DeleteIngress(ctx context.Context, client ssh.Client, names *KubeNames, appInst *edgeproto.AppInst) error {
+	kconfArg := names.GetTenantKconfArg()
+	ingressFile := getIngressManifestName(names)
+	cmd := fmt.Sprintf("kubectl %s delete -f %s", kconfArg, ingressFile)
+	out, err := client.Output(cmd)
+	if err != nil && !strings.Contains(out, "not found") {
+		return fmt.Errorf("failed to delete ingress for %s: %s, %s", ingressFile, out, err)
+	}
+	return nil
+}
+
+func getIngressManifestName(names *KubeNames) string {
+	ingressFile := getIngressFileName(names)
+	return ingressFile
+}
+
+type ingressItems struct {
+	Items []networkingv1.Ingress `json:"items"`
+}
+
+func GetIngresses(ctx context.Context, client ssh.Client, names *KconfNames, ops ...GetObjectsOp) ([]networkingv1.Ingress, error) {
+	data := &ingressItems{}
+	err := GetObjects(ctx, client, names, "ingress", data, ops...)
+	if err != nil {
+		return nil, err
+	}
+	return data.Items, nil
+}
+
+func GetIngressExternalIP(ctx context.Context, client ssh.Client, names *KubeNames, name string) (string, error) {
+	log.SpanLog(ctx, log.DebugLevelInfra, "get ingress IP", "kconf", names.KconfName)
+	for i := 0; i < 60; i++ {
```

Comment:
nit - can you make the wait loop max retries and wait time constants, similar to what was done in `ingress-nginx.go`?

Reply:
makes sense